pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/LLCSegmentCompletionHandlers.java

```diff
@@ -276,10 +307,193 @@ public String segmentUpload(@QueryParam(SegmentCompletionProtocol.PARAM_INSTANCE
+    String tempSegmentDataDirStr = StringUtil.join("/", baseDirStr, segmentNameStr +
+            SEGMENT_TMP_DIR + String.valueOf(System.currentTimeMillis()));
+    File tempSegmentDataDir = new File(tempSegmentDataDirStr);
+    File segDstFile = new File(StringUtil.join("/", tempSegmentDataDirStr, segmentNameStr));
+    // Use PinotFS to copy the segment file to local fs for metadata extraction.
+    PinotFS pinotFS = PinotFSFactory.create(ControllerConf.getUriFromPath(_controllerConf.getDataDir()).getScheme());
+    try {
+      Preconditions.checkState(tempSegmentDataDir.mkdirs(), "Failed to create directory: %s", tempSegmentDataDir);
+      pinotFS.copyToLocalFile(segmentLocation, segDstFile);
+      return getSegmentMetadataFromLocalFile(segmentName, segDstFile);
+    } catch (Exception e) {
+      LOGGER.error("Exception copy segment file to local " + segmentNameStr, e);
+      return null;
+    } finally {
+      FileUtils.deleteQuietly(tempSegmentDataDir);
+    }
+  }
+
+  private SegmentMetadataImpl getSegmentMetadataFromLocalFile(LLCSegmentName segmentName, File segmentFile) {
+    String baseDirStr = StringUtil.join("/", _controllerConf.getDataDir(), segmentName.getTableName());
+    String tempMetadataDirStr = StringUtil.join("/", baseDirStr,
+            segmentName.getSegmentName() + METADATA_TEMP_DIR_SUFFIX + String.valueOf(System.currentTimeMillis()));
+    File tempMetadataDir = new File(tempMetadataDirStr);
+    try (// Extract metadata.properties
+         InputStream metadataPropertiesInputStream = TarGzCompressionUtils
+                 .unTarOneFile(new FileInputStream(segmentFile), V1Constants.MetadataKeys.METADATA_FILE_NAME);
+         // Extract creation.meta
+         InputStream creationMetaInputStream = TarGzCompressionUtils
+                 .unTarOneFile(new FileInputStream(segmentFile), V1Constants.SEGMENT_CREATION_META)
+         ) {
+      Preconditions.checkState(tempMetadataDir.mkdirs(), "Failed to create directory: %s", tempMetadataDirStr);
+      Preconditions.checkNotNull(metadataPropertiesInputStream, "%s does not exist",
+              V1Constants.MetadataKeys.METADATA_FILE_NAME);
+      java.nio.file.Path metadataPropertiesPath =
+              FileSystems.getDefault().getPath(tempMetadataDirStr, V1Constants.MetadataKeys.METADATA_FILE_NAME);
+      Files.copy(metadataPropertiesInputStream, metadataPropertiesPath);
+
+      Preconditions.checkNotNull(creationMetaInputStream, "%s does not exist", V1Constants.SEGMENT_CREATION_META);
+      java.nio.file.Path creationMetaPath = FileSystems.getDefault().getPath(tempMetadataDirStr, V1Constants.SEGMENT_CREATION_META);
+      Files.copy(creationMetaInputStream, creationMetaPath);
+      // Load segment metadata
+      return new SegmentMetadataImpl(tempMetadataDir);
+    } catch (Exception e) {
+      LOGGER.error("Exception extracting and reading segment metadata for " + segmentName.getSegmentName(), e);
```

Comment:
Please use parameters for logging not concatenation

Reply:
Done.